trunk/src/hpinterp/update_gdof.F90

```diff
@@ -437,3 +440,529 @@ subroutine update_gdof()
+!
+!..compute global count and offset
+   j_off = 0
+   if (RANK > 0) j_off = sum(nod_cnt(1:RANK))
+   j_glb = sum(nod_cnt(1:NUM_PROCS))
+!
+!..exit if nobody needs DOFs from other subdomains
+   if (j_glb .eq. 0) goto 70
+!
+!..fill global node list with local nodes
+   allocate(nod_glb(j_glb)); nod_glb = 0
+   if (nod_flg) then
+      nod_glb(j_off+1:j_off+j_loc) = nod_loc(1:j_loc)
+      deallocate(nod_loc)
+   endif
+!
+!..collect local node list from every processor
+   count = j_glb
+   call MPI_ALLREDUCE(MPI_IN_PLACE,nod_glb,count,MPI_INTEGER,MPI_SUM,MPI_COMM_WORLD,ierr)
+!
+!..iterate through the node list and check whether one can supply node data
+   allocate(nod_rnk(j_glb)); nod_rnk = NUM_PROCS
+   do i=1,j_glb
+      nod = nod_glb(i)
+      if (NODES(nod)%visit .eq. 1) nod_rnk(i) = RANK
+   enddo
+!
+!..collect node supplier list
+   count = j_glb
+   call MPI_ALLREDUCE(MPI_IN_PLACE,nod_rnk,count,MPI_INTEGER,MPI_MIN,MPI_COMM_WORLD,ierr)
+!
+!..iterate through node supplier list, and exchange node data
+   rcv = 0; k = nod_cnt(rcv+1)
+   do i=1,j_glb
+!  ...check whether there is no supplier
+      if (nod_rnk(i) .eq. NUM_PROCS) cycle
+!  ...determine supplier, receiver, and dof count
+      src = nod_rnk(i)
+      tag = mod(nod_glb(i),2000000)
+      do while (i > k)
+         rcv = rcv+1
+         k = k+nod_cnt(rcv+1)
+      enddo
+      nod = nod_glb(i)
+      call find_ndof(nod, ndofH,ndofE,ndofV,ndofQ)
+      if (ndofH .eq. 0) then
+         NODES(nod)%visit = 1
+         cycle
+      endif
+      count = 3*ndofH
+!  ...check whether supplier
+      if (src .eq. RANK) then
+         if (.not. associated(NODES(nod)%dof)) then
+            write(*,6010) RANK,'SRC: DOF NOT ASSOCIATED  ',nod,S_Type(NODES(nod)%ntype)
+            stop
+         endif
+         if (.not. associated(NODES(nod)%dof%coord)) then
+            write(*,6010) RANK,'SRC: COORD NOT ASSOCIATED  ',nod,S_Type(NODES(nod)%ntype)
+            stop
+         endif
+         !write(*,6010) RANK,'SENDING  ',nod,S_Type(NODES(nod)%ntype)
+         buf => NODES(nod)%dof%coord
+         call MPI_SEND(buf,count,MPI_REAL8,rcv,tag,MPI_COMM_WORLD,ierr)
+!  ...check whether receiver
+      elseif (rcv .eq. RANK) then
+         if (.not. associated(NODES(nod)%dof)) then
+            write(*,6010) RANK,'RCV: DOF NOT ASSOCIATED  ',nod,S_Type(NODES(nod)%ntype)
+            stop
+         endif
+         if (.not. associated(NODES(nod)%dof%coord)) then
+            write(*,6010) RANK,'RCV: COORD NOT ASSOCIATED  ',nod,S_Type(NODES(nod)%ntype)
+            stop
+         endif
+         !write(*,6010) RANK,'RECEIVING',nod,S_Type(NODES(nod)%ntype)
+         buf => NODES(nod)%dof%coord
+         call MPI_RECV(buf,count,MPI_REAL8,src,tag,MPI_COMM_WORLD,MPI_STATUS_IGNORE,ierr)
+         NODES(nod)%visit = 1
+      endif
+ 6010 format('update_gdof: [',I4,'] ',A,' nod=',I6,', type=',A)
+   enddo
+!
+   deallocate(nod_rnk,nod_glb)
```

Comment:
I think you might have missed doing the resetting of visitation flags for nodes that were not supplied in the communication routines (setting `%visit` from `-1` to `0`)?